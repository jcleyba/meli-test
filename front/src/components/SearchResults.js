import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import queryString from 'query-string';
import Helmet from "react-helmet";

import { fetchItems } from '../actions/searchActions';
import ListContainer from './common/List';
import Spinner from './common/Spinner';
import Breadcrumbs from './common/Breadcrumbs';

const ITEMS_TO_FETCH = 4;

export class SearchResults extends Component {

  constructor(props) {
    super(props);
    this.term = queryString.parse(this.props.location.search).search;
  }

  componentWillMount = () => {
    if (!this.props.items.length) {
      this.props.fetchItems(this.term, ITEMS_TO_FETCH);
    }
  };

  renderMetaTitle = () => {
    // Change Meta title for SEO
    return <Helmet title={`${this.term} - ${this.term} en Mercado Libre Argentina`}/>
  }

  renderResults = () => {
    const {isFetching, items} = this.props;

    if (isFetching) {
      return <Spinner/>
    }
    if (!items.length) return <h4>No hay resultados para su busqueda</h4>;
    return <ListContainer products={items}/>
  };

  render() {
    const pathFromRoot = this.props.filters.length && this.props.filters[0].values[0].path_from_root || [];

    return (
      <div className="search-results">
        {this.renderMetaTitle()}
        <Breadcrumbs pathFromRoot={pathFromRoot}/>
        {this.props.error}
        {this.renderResults()}
      </div>
    );
  }
}

const mapStateToProps = ({searchReducer}) => {
  const {items, error, filters, isFetching} = searchReducer;
  return {items, error, isFetching, filters}
};

const routedComponent = withRouter(SearchResults);
export default connect(mapStateToProps, {fetchItems})(routedComponent);
