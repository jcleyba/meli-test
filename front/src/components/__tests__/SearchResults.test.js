import React from 'react';
import { shallow, mount } from 'enzyme';
import 'jest-enzyme';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';

import { SearchResults } from '../SearchResults';

describe('ProductDetail Component', () => {

  let wrapper;
  let mounted;

  const props = {
    items: [{
      id: 1,
      title: 'Title',
      price: {amount: 10},
      address: {state_name: 'CÃ³rdoba'}
    }],
    isFetching: false,
    filters: [],
    location: {
      search: ''
    }
  };

  const Component = (props) => {
    return <MemoryRouter>
      <SearchResults {...props}/>
    </MemoryRouter>
  };


  beforeEach(() => {
    wrapper = shallow(<Component {...props}/>);
    mounted = mount(<Component {...props}/>);
  });

  afterEach(() => {
    wrapper = null;
    mounted = null;
  });

  describe('SearchResults', () => {

    it('matches current snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should have isFetching prop', () => {
      expect(mounted.find('SearchResults').props()).toHaveProperty('isFetching');
    });

    it('should have items prop', () => {
      expect(mounted.find('SearchResults').props()).toHaveProperty('items');
      expect(mounted.find('SearchResults').props().items).toHaveLength(1);
    });

    it('should have a search-results className present', () => {
      expect(mounted.find('.search-results')).toBePresent();
    });

    it('should contain a List Container', () => {
      expect(mounted.find('ListContainer')).toBePresent();
    });

    it('should contain a searching text', () => {
      mounted = mount(<Component {...props} isFetching={true}/>);

      expect(mounted.find('.loading')).toBePresent();
    });
  })

})
;
