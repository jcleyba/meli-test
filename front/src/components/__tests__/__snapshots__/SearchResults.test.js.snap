// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail Component SearchResults matches current snapshot 1`] = `"<div class=\\"search-results\\"><ol class=\\"breadcrumbs\\"></ol><div class=\\"list-container\\"><div class=\\"list-item\\"><div class=\\"thumbnail\\"><a aria-current=\\"false\\" href=\\"/items/1\\"><img alt=\\"Title\\" title=\\"Title\\" class=\\"product-thumb\\"/></a></div><div class=\\"description\\"><div class=\\"price\\"><span class=\\"amount\\">$ 10</span></div><div class=\\"title\\"><a aria-current=\\"false\\" href=\\"/items/1\\"> Title</a></div></div><div class=\\"location\\">CÃ³rdoba</div></div></div></div>"`;
